### PR TITLE
Added proxy, is_focused and patched driver for latest google version

### PR DESCRIPTION
**Pull request by ManiMozaffar**
Added support for user pass proxy

Patched driver with latest google chrome, so it can bypass websites as mentioned in issues

### CODE REVIEW

**Comment by mdmintz** [comment]
It looks like this uses the same modified undetected-chromedriver code that I have here: https://github.com/seleniumbase/SeleniumBase/tree/master/seleniumbase/undetected
which includes some of the patches made by @lukect and @fabifont here: https://github.com/ultrafunkamsterdam/undetected-chromedriver/pull/1010 and the proxy code I have here: https://github.com/seleniumbase/SeleniumBase/blob/master/seleniumbase/core/proxy_helper.py

This can also be achieved through seleniumbase directly: (``pip install seleniumbase``):

The available args can be found here: https://github.com/seleniumbase/SeleniumBase/blob/master/seleniumbase/plugins/driver_manager.py

Simple example:

```python
from seleniumbase import Driver
driver = Driver(uc=True)
```

To add in a proxy, here are the available args with default values:

```python
proxy=None,  # Use proxy. Format: "SERVER:PORT" or "USER:PASS@SERVER:PORT".
proxy_bypass_list=None,  # Skip proxy when using the listed domains.
proxy_pac_url=None,  # Use PAC file. (Format: URL or USERNAME:PASSWORD@URL)
```

**Comment by ManiMozaffar** [comment]
> It looks like this uses the same modified undetected-chromedriver code that I have here: https://github.com/seleniumbase/SeleniumBase/tree/master/seleniumbase/undetected which includes some of the patches made by @lukect and @fabifont here: #1010 and the proxy code I have here: https://github.com/seleniumbase/SeleniumBase/blob/master/seleniumbase/core/proxy_helper.py
> 
> This can also be achieved through seleniumbase directly: (`pip install seleniumbase`):
> 
> The available args can be found here: https://github.com/seleniumbase/SeleniumBase/blob/master/seleniumbase/plugins/driver_manager.py
> 
> Simple example:
> 
> ```python
> from seleniumbase import Driver
> driver = Driver(uc=True)
> ```

Indeed, for most parts I merged them into one as @ultrafunkamsterdam has been inactive recently.
I just saw your proxy code, the proxy also contains IP in file creation without zipping, as if you run multiple chrome together it may causes some error if they try to access same extensions over and over, and overwriting it with new host, port, password and user.

**Comment by mdmintz** [comment]
> @ultrafunkamsterdam has been inactive recently.

He's definitely around, but possibly busy. He assigned a PR to himself yesterday:

Screenshot 2023-02-04 at 11 44 11 AM

As for the Chrome multi-threading, that's why the file-locking is used. But if the different Chrome browsers require different proxies, then different zip files are needed too.

**Comment by lukect** [comment]
You should only fix one thing per PR so it is more manageable for maintainers.

**Comment by lukect** [comment]
Proxy feature is definitely needed. 👍
You have definitely improved the download code. 👍

**Comment by ManiMozaffar** [comment]
@ultrafunkamsterdam 
Hey leon, I have resolved the conflicts.

**Comment by Carlos-OL** [comment]
I am testing this code that works with the normal selenium webdriver, but when I try it with UC, I get an error. Something is wrong with "desired capabilities" 

`
proxy_ip = "XXXX:YYYY
proxy = Proxy()
proxy.proxy_type = ProxyType.MANUAL
proxy.http_proxy = proxy_ip
proxy.ssl_proxy = proxy_ip

capabilities = webdriver.DesiredCapabilities.CHROME
proxy.add_to_capabilities(capabilities)

options = webdriver.ChromeOptions()
options.add_experimental_option('detach', True)

driver = uc.Chrome(desired_capabilities=capabilities, options=options)`

**Comment by ManiMozaffar** [comment]
@Carlos-OL 
That does not support user and password HTTP proxy, however my solution works with that as well.

**Comment by Carlos-OL** [comment]
> @Carlos-OL That does not support user and password HTTP proxy, however my solution works with that as well.

Hey, thanks for responding. I am not sure what you mean. My post above works fine with selenium webdriver, so why wouldn't it work with UC?
Also, my proxy is IP:PORT (that is what I meant with XXXX:YYYY). It is not user:pass.

When you say you added support for "proxy", how exactly should I enter the option?
I tried the code above and it didn't work.  I also tried:
`driver = uc.Chrome(proxy="IP:PORT")`

and that didnt work either. Perhaps I am missing the right notation? Can you please help me??
Thanks

**Comment by sufianadnan** [comment]
How would i use this, i currently have the original UC, (im not really a big github person), how would i setup using this, anything helps

**Comment by ManiMozaffar** [comment]
@sufianadnan @Carlos-OL 
py -m pip install git+https://github.com/ManiMozaffar/undetected-chromedriver/tree/beta

**Comment by Carlos-OL** [comment]
> @sufianadnan @Carlos-OL
> 
> ```
> py -m pip install git+https://github.com/ManiMozaffar/undetected-chromedriver/tree/beta
> ```
> 
> after using this, you may connect your proxy to selenium with:
> 
> ```python
> driver = uc.Chrome(proxy="IP:PORT")
> #or   
> driver = uc.Chrome(proxy="USER:PASSWORD@IP:PORT")
> ```

@ManiMozaffar I tried the link you provided above and I got an error that, that repository does not exist anymore. Did you move/deleted it? I also tried the link in your page, but that seemed to be an older version 3.2.1 and did not contain the proxy file. Can you please advise?
Thanks

**Comment by Carlos-OL** [comment]
@ManiMozaffar I was finally able to install your beta branch via zip file. No idea why no other method worked.
The proxy feature is not working for me. Here is the error I get when trying to open any website:
ERR_NO_SUPPORTED_PROXIES
My proxy is host:port and it works on the latest selenium version, and I have tested it with 3rd party extensions.
image

**Comment by rkt231** [comment]
Hi,

I am wondering if the old method still works:

```
import undetected_chromedriver as UC
_opts = UC.ChromeOptions()
_opts.add_argument('--proxy-server=%s' % conf.PROXIES['https'])
```

?

That would be helpful (to be more compatible with selenium{wire}),

Thanks,

File: undetected_chromedriver/__init__.py [hunk 1]
```diff
@@ -279,6 +298,29 @@ def __init__(
...
+                    user_password, ip_port = proxy.split("@")
+                    user, password = user_password.split(':')
+                    ip, port = ip_port.split(':') 
+                    proxy_extension = ProxyExtension(ip, int(port), user, password)
+                    options.add_argument(f"--load-extension={proxy_extension.directory}")
+                else:
+                    ip, port = proxy.split(':') 
+                
+                port = int(port)
+                options.add_argument(f"--proxy-server=%s' % '{ip}:{int(port)}")
```
Comment (BaiShaojie123123) [inline undetected_chromedriver/__init__.py:319]: The above code has a username password, there is no password here, maybe it should be indented in "else" here?

**Comment by ManiMozaffar** [comment]
> Hi,
> 
> I am wondering if the old method still works:
> 
> ```
> import undetected_chromedriver as UC
> _opts = UC.ChromeOptions()
> _opts.add_argument('--proxy-server=%s' % conf.PROXIES['https'])
> ```
> 
> ?
> 
> That would be helpful (to be more compatible with selenium{wire}),
> 
> Thanks,


> The above code has a username password, there is no password here, maybe it should be indented in "else" here?


Hello, 
That argument does not support user password proxy.  so an extension is used connect to proxies with user password if it has.
Although selenium wire uses a different approach, starts a proxy server on your local machine then automatically configures the browser to use the proxy server and as you navigate to a web page, the browser sends an HTTP request to the web server via the proxy server. Then The proxy server intercepts the request and forwards it to the web server. In other words,  there is man in the middle, so it doesn't matter how and which proxy to use to send the request, but detection-wise using it is not a good practice since SSL fingerprint are detectable because of TLS handshake messages exchanged between the client and server. To avoid being detected you should use custom SSL certificate for HTTPS requests, however, using a custom SSL certificate may also trigger other detection measures on the server side, such as certificate pinning or certificate transparency.
I think using extension is best solution, that's why I added it like that. but the owner of repo has not left me a comment on this pull request yet.



> ManiMozaffar I was finally able to install your beta branch via zip file. No idea why no other method worked.
> The proxy feature is not working for me. Here is the error I get when trying to open any website:
> ERR_NO_SUPPORTED_PROXIES

Please share me your code, it should be working. Note that you should install my repo if you want to use it. It's been old comparing to latest version of UC and I contin> ERR_NO_SUPPORTED_PROXIES

Please share me your code, it should be working. Note that you should install my repo if you want to use it. It's been old comparing to latest version of UC and I continued my own version which I'm not going to pull it here. used my own version which I'm not going to PR it here.  Make sure you have uninstalled the last version of UC. 
```
pip uninstall undetected-chromedriver
pip install git+https://github.com/ManiMozaffar/undetected-chromedriver.git@beta
```

**Comment by rkt231** [comment]
Thanks for the clarification @ManiMozaffar  ; it is crystal clear.

**Comment by nugtheplug** [comment]
Hi, I am trying to use your fork so that I can execute tests with USER:PASS proxies - however, I am getting this error when running my code: 

File "/lib/python3.10/site-packages/undetected_chromedriver/__init__.py", line 318, in __init__
    port = int(port)
ValueError: invalid literal for int() with base 10: 'PORT'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/v/Desktop/Python/test/test/lib/python3.10/site-packages/undetected_chromedriver/__init__.py", line 313, in __init__
    proxy_extension = ProxyExtension(ip, int(port), user, password)
NameError: name 'ProxyExtension' is not defined. Did you mean: 'proxy_extension'?

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/v/Desktop/Python/test/modules/main.py", line 261, in <module>
   main()
  File "/Users/v/Desktop/Python/test/modules/main.py", line 255, in main
    driver = uc.Chrome(proxy="usernamexxx:passwordxxx@IPxxx:Portxxx")
  File "/Users/v/Desktop/Python/test/test/lib/python3.10/site-packages/undetected_chromedriver/__init__.py", line 321, in __init__
    raise ValueError("Proxy Configuration is failed, please make sure the input is username:password@ip, port")
ValueError: Proxy Configuration is failed, please make sure the input is username:password@ip, port
Exception ignored in: <function Chrome.__del__ at 0x1110a65f0>
Traceback (most recent call last):
  File "/Users/v/Desktop/Python/test/test/lib/python3.10/site-packages/undetected_chromedriver/__init__.py", line 833, in __del__
    self.quit()
  File "/Users/v/Desktop/Python/test/test/lib/python3.10/site-packages/undetected_chromedriver/__init__.py", line 789, in quit
    self.service.stop()
  File "/Users/v/Desktop/Python/test/test/lib/python3.10/site-packages/undetected_chromedriver/__init__.py", line 796, in __getattribute__
    return super().__getattribute__(item)
AttributeError: 'Chrome' object has no attribute 'service'

--------------------------

My code is only: 

def main():
        driver = uc.Chrome(proxy=proxy) #its a user:pass proxy like I mentioned
        driver.get('https://httpbin.org/ip')
        #sleep so I can check its using the right IP ect

if __name__ == '__main__':
   main()
   
   -------------------------
   
 Sorry if this is obvious, or my comment is not formatted properly this is the first time I have ever posted/asked something on Github
 
 Thank you in advance

**Pull request closed without merging**